### ISSUE TITLE
Request: Implement GeoPackage

### ISSUE THREAD

**Issue by hexamon** [2shady4u/godot-sqlite#18]
### **What is it?**

GeoPackage is a spatial extension of SQLite. This allows for storing and querying data in a spatially intelligent way. It is used by GIS (Geographical Information System) software packages for a wide variety of purposes. Things that can be stored are for example building geometries, aerial photography, points of interest, land use, etc etc.

[GeoPackage.org](https://www.geopackage.org/)

### Why implement it in Godot?

A game engine is spatial in nature. For projects that deal with a large environment GeoPackage could be very useful due its built in spatial logic. You can execute queries such as:

- Give me all rows from table X that are within 500 world units of a given point.
- Give me all rows from table X that overlap something in table Y.
- Give me all rows from table X that overlap a 500 world unit bufferzone around point Y.

If you are building a large world you can for example use this to stream assets by keeping their world location data in a GeoPackage.

It also supports storing and tiling of large raster datasets. These rasters could be used as textures or to map other assets into Godot.

Perhaps most importantly, it also allows for spatial indexing. So you can retrieve rows based on their location very quickly from a large dataset (eg. load all the houses within 500 world units from the player.)

Since there exists a large number of free and open source software packages (such as [QGIS ](https://qgis.org/en/site/) for example) for working with spatial data, interesting synergies could arise as  well. You could for example use hydrological models or routing analysis directly on data that you read with Godot. QGIS could also function as a de facto map editor in a Godot project that stores its world data in GeoPackage.

Thank you for your time and your work on implementing SQLite in Godot!

**Comment by 2shady4u** [comment]
Well.. I'm for sure interested in such a format, It's just that I don't really know if:
A. I will have time to implement such a thing 🤔 I currently am a bit short on time 😖 
B. If all gdsqlite users will have need of a geopackage implementation? Maybe this might be better done in a different repo with a different target audience? I mean, I could fork godot-sqlite and do this if this is the way to do 🤔 

A lot of maybes here 🤔 
Do you have a personal use-case in mind?
I'll have to ponder on this one a bit...

**Comment by hexamon** [comment]
Thanks for your reply!

As for...
A. Totally understand that. I mainly wanted to write this down so it's there if you or anyone else has time and interest.
B. Most users probably wont have a use for GeoPackage functionality. However, it would in my view be extremely powerful in a select set of cases.

The main point would be to allow for large, hand crafted open worlds. Since GeoPackage is basically SQLite with spatial awareness, you would in effect have an extra "meta coordinate system" if you will.

The main issue with large open worlds is of course floating point precision problems and the amount of simultaneous assets in the scene. So as you probably know you usually get around this by moving the world origin as the player moves around and load/unload assets by transforming between a global and scene coordinate system (or using procgen or something).

But if you want a detailed, large world with lots of (maybe exclusively) hand placed assets, finding and loading the right assets in a timely manner can get tricky.

Here's where GeoPackage would come in. For example, you could store your scenery in a table with the following structure:

```
geometry | assetPath
-----------------------
 a point | res://asset
```

Let's then say your viewdistance is 1000 world units. Then you could continually check for all assets within the player's viewdistance plus some more to make sure they're loaded:

```
SELECT 
  assetPath 
FROM 
  scenery 
WHERE 
  st_distance(player_global_position,geometry) >= 1200
```

And just flip that around to unload assets which are no longer close to view distance.
The main strength with GeoPackage here is the spatial SQL and the spatial index. You should be able to run a query like that on millions of rows fairly quickly. Then you could have a huge, hand crafted world (made in a GIS software like QGIS for example) that you easily stream in and out of Godot.

You can also process geometry in GeoPackage with SQL. Let's say you have a very large polygon (10 000 x 10 000 world units) that represents a forest, and you want to place trees in it using noise. You could clip the polygon with the player's view distance directly with SQL before loading the vertices in Godot by making a circular buffer polygon around the player and intersecting that with the forest polygon:

```
SELECT
  st_intersect(st_buffer(player_global_position,viewDistance),geometry) as geometry
FROM
  forest
WHERE
  st_distance(player_global_position,geometry) >= viewDistance
```

To sum up, since GeoPackage is made to handle real world spatial data it's a nice fit for handling any type of large world data with precision. Which brings me to another important point; that a native GeoPackage reader in Godot means a native reader for real world geodata as well. If this was implemented you could easily read OpenStreetMap data, terrain data from NASA, satellite photos etc etc.

**Comment by 2shady4u** [comment]
Very interesting 🤔 

Just a small question: I would expect these assets to be stored in the SQLite databases as BLOBs for maximum efficiency? 
Is this correct? And if yes, I'm still in the (slow) process of implementing support for BLOBs in the bindings. (see the [blob-testing](https://github.com/2shady4u/godot-sqlite/tree/blob-testing) -branch)

Second small question: How adept are you in using the C++ bindings of Geopackage?

**Comment by hexamon** [comment]
Correct, the contents of the geometry column is stored as BLOB.

[GeoPackage spec](http://www.geopackage.org/spec/):

_Geometry encoded as per clause Geometry Encoding. <geometry type_name> is one of the core geometry types listed in Geometry Types (Normative) encoded per clause 2.1.3 or a geometry type encoded per an extension such as GeoPackage Non-Linear Geometry Types. Geometry Types XY, XYZ, XYM and XYZM geometries use the same data type. **Stored as SQLite BLOB**._

I have zero experience with C++ sadly. I work with geodata/spatial analysis in my dayjob so I understand a lot of the stuff at an abstract level, but I'm mostly a hobbyist when it comes to the nuts and bolts of the softwares.

One of the most prolific FOSS packages around in the geospatial world is GDAL. It is coded in C/C++ and has a [driver implemented for GeoPackage.](https://github.com/OSGeo/gdal/blob/47b041c8cc2e0ed871e868615163facc56d6c2f8/gdal/doc/source/drivers/vector/gpkg.rst) Might be worth checking out.

However! It struck me that there is another spatial SQLite variant; [Spatialite.](https://www.gaia-gis.it/fossil/libspatialite/index) It was developed before GeoPackage and I assumed GeoPackage was just the same thing but more widely adopted. But now that I've read a bit more about it, it seems Spatialite might be a better fit for Godot. GeoPackage is more of a transfer format and Spatialite aims for a large spatial SQL toolset for manipulating data (ie. what I would want from an implementation in Godot.) Introducing yet another spatial SQLite variant might just confuse things, but I think implementing GeoPackage or Spatialite would be incredible if it happened.

**Comment by 2shady4u** [comment]
Dropping this here for later reference:
https://github.com/boku-ilen/geodot-plugin

**Comment by MathiasBaumgartinger** [comment]
Now that I saw this Issue i just wanted to check in real quick and mention that we provide an addon called geodot which is able to load geodata from a geopackage (as @2shady4u mentioned). If you have any specific questions feel free to join our server at https://discord.gg/ar4xPTqa and ask questions or discuss new features, or simply by adding issues/pull requests in the repo. It is mostly my colleague @kb173 who maintains this repository, I think we are both gladly available for you though.

**Comment by 2shady4u** [comment]
Closing this issue because of two main reasons:
- Implementation of GeoPackage is not in this project's current scope or roadmap.
- There's other solutions that successfully implement GeoPackage (as found [here](https://github.com/boku-ilen/geodot-plugin))

Please feel free to open another issue if you have other issues or suggestions that are unrelated to geodata! 😄
